### DISCUSSION TITLE
How do I convert the recieved metadata to string so that I can show it on an SPI TFT display.

### DISCUSSION THREAD

**Discussion by msaraswat15031997** [pschatzmann/ESP32-A2DP#383] [Q&A] [answered]
Hello sir, I am building a bluetooth speaker with a TFT display, I have an i2s amp connected and using A2DP sink , I want to show the metadata recieved  on the tft but I don't know how to convert it to a string. I get the title, and time on the serial monitor but  Also want to show things like the volume , artist, title , time etc and whatever possible but I don't know how to. So sorry I am not an expert in programming.

**Comment by pschatzmann** [comment, marked as answer]
Why don't just just look at the [provided example](https://github.com/pschatzmann/ESP32-A2DP/blob/main/examples/bt_music_receiver_with_metadata/bt_music_receiver_with_metadata.ino) and try it out?
Your question does not make any sense because you can see that you are actually getting strings...

To decode the id you can use https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/bluetooth/esp_avrc.html#_CPPv423esp_avrc_md_attr_mask_t

**Comment by msaraswat15031997** [comment]
I'm really Sorrry sir. Here's my code, i need to print the song title on the TFT in the loop funciton.

#include <TFT_eSPI.h>                 
#include <SPI.h>
TFT_eSPI tft = TFT_eSPI();

#include "BluetoothA2DPSink.h"
BluetoothA2DPSink a2dp_sink;
bool is_active = true;

void setup(void) {
  tft.init();
  tft.setRotation(1);


i2s_pin_config_t my_pin_config = 
  {
   .bck_io_num = 26,
   .ws_io_num = 25,
   .data_out_num = 33,
   .data_in_num = I2S_PIN_NO_CHANGE
   };
 a2dp_sink.set_pin_config(my_pin_config);
 a2dp_sink.set_avrc_metadata_callback(avrc_metadata_callback);
 a2dp_sink.set_avrc_metadata_attribute_mask(ESP_AVRC_MD_ATTR_TITLE | ESP_AVRC_MD_ATTR_PLAYING_TIME |ESP_AVRC_MD_ATTR_ARTIST );
 a2dp_sink.start("bt");
}

void avrc_metadata_callback(uint8_t data1, const uint8_t *data2) {
 Serial.printf("AVRC metadata rsp: attribute id 0x%x, %s\n", data1, data2);
}

void loop() {
  tft.fillScreen(TFT_GREY);
tft.setCursor(0, 0, 2);
tft.setTextColor(TFT_WHITE,TFT_BLACK);  tft.setTextSize(1);
tft.println(songname);
}

**Reply by pschatzmann** [reply]
Why don't you just move the output from the loop to the callback ? And make sure that you select the titles only with data == ESP_AVRC_MD_ATTR_TITLE.
If this has a negative effect on the audio you would just need to save the __a copy of the title__ when you get it in a global variable.

**Reply by msaraswat15031997** [reply]
So I create a 

Char xyz = ESP_AVRC_MD_ATTR_TITLE; 

Will this put the title info in xyz? 


I'm so sorry to have caused trouble.

**Comment by pschatzmann** [comment]
I would try the following first:
```C++

void avrc_metadata_callback(uint8_t typ, const uint8_t *text) {
  if (typ== ESP_AVRC_MD_ATTR_TITLE){
    tft.fillScreen(TFT_GREY);
    tft.setCursor(0, 0, 2);
    tft.setTextColor(TFT_WHITE,TFT_BLACK); 
    tft.setTextSize(1);
    tft.println(text);
  }
}

void loop() {
}

```

**Reply by msaraswat15031997** [reply]
Wow,  you're a savior,  thank you for such quick replies.  Running right now and trying it out  :D

**Reply by msaraswat15031997** [reply]
![20230312_213737.jpg](https://user-images.githubusercontent.com/53805706/224557230-4c8de93e-9b67-416e-8b05-68d4e2455447.jpg)

Just WOW! I got the text up and running on the tft. 

Also how do I display the remaining data,like the time and artist, ?
